### PR TITLE
MRG, MAINT: Refactor volume interpolation code

### PR DESCRIPTION
**Pull request by larsoner**
I am still annoyed by how slow the volumetric morphing is, so I decided to see if I could alleviate one bottleneck, namely the linear and zero-order interpolations. I never 100% understood Matti's volume interpolation code I adapted so many years ago, but based on how it works it should definitely be faster when we have lots of volumes to morph. Whether it's already the same speed as dipy and nibabel (scipy.ndimage) or faster with 1, 10, 100, or 1000 volumes remains to be determined, but I wouldn't be surprised if it were about the same speed with even just 1. It definitely will be with 1000.

In any case, to move toward being able to use the sparse matrix code I:

1. Refactored our `_add_interpolator` code to do a generic grid-to-grid interpolation
2. Added tests showing equivalences between `nibabel`, `dipy`, and our code for linear and nearest interpolation (it's a ton of combos but it takes only 0.3 sec on my machine)
3. Made the code `jit`-able but it didn't actually speed it up (thanks, vectorization!)
4. Fixed some minor bugs with our treatment of edge/border volume vertices in `_add_interpolator` (exposed by the nibabel/dipy equivalency tests), where some vertices could be interpolated to by a subset of the 8 possible vertices (e.g., 4 or 2 or even 1 of them) but it only added the interpolation if all could.

This should allow me to then speed up `SourceMorph._morph_vols`, but I'll do that in a subsequent PR. Even if it does not speed things up, this is already a nice refactoring + unit testing of our code.

### CODE REVIEW

File: mne/source_space.py [hunk 1]
```diff
@@ -2302,109 +2302,140 @@ def _add_interpolator(sp):
...
+
+    # Compose the sparse matrices
+    for si, s in enumerate(sp):
+        # limit it columns that have any contribution from inuse
+        any_ = sparse.diags(
+            np.asarray(
+                interp[:, s['inuse'].astype(bool)].sum(1)
+            )[:, 0].astype(bool).astype(float)
+        )
+        s['interpolator'] = any_ * interp
```
Comment (agramfort) [inline mne/source_space.py:2321]:
```suggestion
        s['interpolator'] = any_ @ interp
```

to be more explicit

**Review by agramfort — COMMENTED** [review]
just a nitpick. 

I don't understand the detail of what you did but I assume you checked carefully that there is no regression or behavior change...

thanks

**Review by agramfort — APPROVED** [review]

**Comment by larsoner** [comment]
Yes this even fixes some minor bugs in master exposed by the new unit tests. And the existing tests were already pretty good at detecting when something went wrong as well.

**Comment by agramfort** [comment]
+1 for MRG then

>

**Pull request merged**